### ISSUE TITLE
DVB not selecting ISDBT

### ISSUE THREAD

**Issue by fvp** [mpv-player/mpv#12218]
### Important Information

Provide following Information:
- mpv version: 0.34.1
- Linux Distribution and Version: Ubuntu 22.04
- Source of the mpv binary: Ubuntu repository

### Reproduction steps

I have a multi standard card TBS5230, which I would like to tune to ISDBT.

I see in the logs that mpv detects that the card support ISDBT:
[   0.020][v][dvbin] Querying tuner frontend type via DVBv5 API for frontend FD 7
[   0.020][v][dvbin] DVBv5: Number of supported delivery systems: 7
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be DVB-T
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be DVB-T2
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be ISDBT
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be DVB-C ANNEX A
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be DVB-C ANNEX B
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be DVB-C ANNEX C
[   0.020][v][dvbin] DVBv5: Tuner frontend type seems to be ATSC

But it doesn't try to tune to ISDBT, I can see in the logs that mpv never tries ISDBT, it tries other delivery systems. It thinks that the channels are DVB-C ANNEX A:

[   0.021][i][dvbin] dvb_tune DVB-C ANNEX A Freq: 473142857
[   0.021][v][dvbin] TUNE_IT, fd_frontend 7, DVB-C ANNEX A freq 473142857, srate 0, pol V, diseqc 0

Is there a way to force to use ISDBT ? or fix the delivery system selector?

### Expected behavior

mpv should work with ISDBT channels

### Actual behavior

mpv never tries to tune to ISDBT channels

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/12402123/output.txt)

**Comment by olifre** [comment]
Thanks for the report!
I've been the one mostly maintaining DVB support, but since I don't have any ISDBT hardware (nor possibility to receive around here), this was in fact never tested / implemented. 

Do you have the chance / ability to recompile / test changes? 

If we are very lucky, it is already sufficient to extend all case conditions, such as the one here:
https://github.com/mpv-player/mpv/blob/af9b53f3a3831531504906882efce0d979506ed3/stream/stream_dvb.c#L1186C8-L1186C8
accordingly. The `SYS_*` which needs to be added is `SYS_ISDBT`. In theory, it should behave similar to `SYS_DVBT` or `SYS_DVBT2`.

**Comment by fvp** [comment]
H Olifre,

Yes I could recompile and test the changes.

I agree with you, it should be similar to DVBT and DVBT2.

Br,
Felipe

**Comment by olifre** [comment]
Hi Felipe,
great, then I will have a go in a feature branch — hope to have something ready for you to test later this week (there are 2-3 code places which are not as straightforward as just extending the code with another `case` for the delivery system). Of course, if you want to play with the code before I get to it, that's also possible ;-). 
Cheers, Oliver

**Comment by fvp** [comment]
Sounds good. Thanks

**Comment by fvp** [comment]
I tried with channels.conf.isdbt, got a new error:
[output.txt](https://github.com/mpv-player/mpv/files/12424460/output.txt)

I created the channels.conf with this command: dvbv5-scan ./ph.txt --output-format=ZAP
[ph.txt](https://github.com/mpv-player/mpv/files/12424462/ph.txt)

**Comment by olifre** [comment]
@fvp Thanks! So we are one step further, the channels list is now parsed, but then it bails out. I'm not fully sure why the DVB configuration is deemed invalid, but the list of PIDs for the channels looks fishy, and PIDs are also not listed in the `ph.txt` you attacked, while `dvbv5-scan` usually includes these (but maybe not for ISDBT). 

Can you also attach the actual `channels.conf.isdbt` you created from that, which should be formatted a bit differently? Was that done manually or with another tool? 

I'll hopefully have more time to dig deeper during the weekend.

**Comment by fvp** [comment]
Sure, here we go:
[channels.conf.txt](https://github.com/mpv-player/mpv/files/12424981/channels.conf.txt)

It does seem the pids are different from the ones in the file.

The file was created using dvbv5-scan, as described in my other post.

Thanks

**Comment by olifre** [comment]
Thanks, that's very helpful indeed! It confirms my suspicion that the `channels.conf` was not converted to VDR format (and that should also not be needed in this case). However, the current code detects it as VDR-style channel configuration. 

With your file, I should be able to check out whythat happens and sort this out ;-).

**Comment by fvp** [comment]
Hi @olifre ,

It seems the format VDR doesn't support ISDBT, I get this error:
WARNING: entry 0: delivery system 8 not supported on this format. skipping entry

I changed the channels.conf, I tried changing only the HIERARCHY_NONE, and also remove all "zeros":
[channels.conf.isdbt.txt](https://github.com/mpv-player/mpv/files/12433146/channels.conf.isdbt.txt)

Both ways I got the same error:
[output.txt](https://github.com/mpv-player/mpv/files/12433147/output.txt)

Perhaps something else is happening, because even when the pids were wrong and it was detecting it as DVB-C, it would not print the error that the DVB configuration file was empty.

Thanks

**Comment by olifre** [comment]
Hi @fvp ,

thanks for the quick response! 
Too bad `v4l-utils` / `dvbv5-scan` can not create VDR format for this delivery system. VDR supports this in general, so it's a bit strange. 

Then the way to go for now (until I come up with an automatic fallback) is indeed to replace the zeros, as you did. However, the log you got is very unexpected. 

I tricked `mpv` into believing my local DVB-S2 card can do ISDBT just now, and fed it your channel configuration, and I end up in the tuning stage. So there is something strange going on indeed. "EMPTY" configuration state should only happen if there is no adapter with any matching channel list (i.e. for the same delivery system) found, but in your case, the channel list seems parsed (correctly this time). 

I have pushed another commit to the branch you are using, which does two things:
* For any card which is found, it's forced to assume the card supports ISDBT (only). This eases testing for me, mainly. 
* Lots of debug output is added which hopefully helps to find out where things go awry. 

It will be interesting to see the new output — I'll probably only come back to this in two days from now, but hopefully this will help :wink: .

**Comment by olifre** [comment]
Hi @fvp ,

no problem, that's great news indeed :+1: . The logs look exactly as I would have expected them to look, so it's quite strange things did not work when I did not force any adapter to be detected as `ISDBT` in the code. 

I have pushed another commit on top of my test branch, dropping this enforcement and adding a bit more debugging. I would expect that this fails again, but hopefully, those logs from the failure will finally reveal something. Could you please git it another try? 

Cheers and thanks!

**Comment by fvp** [comment]
Hi @olifre ,

With this new version the issue came back:
[output.txt](https://github.com/mpv-player/mpv/files/12468145/output.txt)

**Comment by olifre** [comment]
Hi @fvp ,

perfect, this is what I expected — and the debug output is helpful! :+1: . 
Actually, this uncovered a bug in channel list creation which could cause the list to be lost if an adapter supported multiple delivery systems and some did not have a config file. As usual, quite astonishing nobody else has hit this yet. You were likely the first since your card supports 7 delivery systems. 

Can you try once more after the very last commit? 
I would expect it to work again, and this time the code does not enforce the adapter to be of ISDBT type :wink: . 

If that does the trick, the only thing left on my todo list is to find a clean way to distinguish between the ISDBT ZAP file created by `dvbv5_scan` (without patching out the zeroes) and a VDR style config file. I'll think about this, likely tomorrow, and ask you for a final test once I have this implemented.

**Comment by fvp** [comment]
Sure, I can try as many time as you need to. Here we go:
[output.txt](https://github.com/mpv-player/mpv/files/12469020/output.txt)

**Comment by olifre** [comment]
Thanks, that looks good :+1:. 

Sadly, the handling of 0's is already doing exactly this, i.e. using default values. The problem is differentiating a DVB-T / ISDBT ZAP file from a VDR config file, since both have 13 columns. Usually, checking which columns contain numbers and which contain text is sufficient to disentangle them, but in the ISDBT case for files created by `dvbv5-scan`, columns which should contain text contain numbers... 

However, I think the easiest approach will be to just check the very first field after the frequency in addition to checking the number and type of columns — this contains the `INVERSION`, which should always be set (at least `dvbv5-scan` and `w_scan` always set this). I'll implement this tomorrow and give the code a final test against my DVB-S2 and DVB-T2 receiver here, clean up the debug output a bit, and when I'm done, I'll ping you for another test :wink: .

**Comment by olifre** [comment]
@fvp I am done :wink:. 

I have pushed my changes into a fresh branch, cleaning up all the intermediate debugging commits:
https://github.com/olifre/mpv/tree/dvb-isdbt-and-cleanup

Could you please give this branch another test? 
First off, it should work and tune just fine. However, could you please also give it a test with the original `channels.conf` with the "0"s as produced by `dvbv5-scan`? If I am correct, this should now also work without problems :smile: .

**Comment by fvp** [comment]
Here we go, with the original channels.conf.isdbt. It is working.
[output.txt](https://github.com/mpv-player/mpv/files/12478986/output.txt)

I also tried without the .isdbt ending. It thinks it is a DVB-C ANNEX A. Tried this just if this could be important for your code:
[output.txt](https://github.com/mpv-player/mpv/files/12478999/output.txt)

Thanks

**Comment by olifre** [comment]
@fvp Many thanks! 
That looks as expected — great to see `channels.conf.isdbt` now works without modifications. 

Indeed, the second output is also kind of expected: `mpv` tried all different possible interpretations of the channels file, i.e. it also tried `ISDBT` along the way, but then it has to choose one way to tune. Sadly, the "ZAP" format is not specific enough — I tried just now to enforce that each line is fully parsed, and while this fixed some situations, it broke others (e.g. `w_scan` produces slightly different channel config files in "ZAP" format than `dvbv5-scan`). What `mpv` currently does it do parse as "ZAP", read as many columns as are there and can be interpreted, and hope for the best if only a `channels.conf` is provided. 

In general, the best way to go would be to use `VDR` style configs, these look the same for any delivery system and the format is well documented. Sadly, it seems there is no channel scanner out there who can produce these for ISDBT. 

So I think the safest route (such as not to break existing channel configs) is to use the specific channel config names to differentiate when non-VDR configs are used, and warn in the docs not to use `channels.conf` but only the specific names whenever possible. 

To prevent future confusion, I have added a note to the docs (which I will include in the pull request including all the DVB fixes we collected here :wink: ).
